### PR TITLE
Modernize test suite

### PR DESCRIPTION
**Pull request by doutriaux1**
@durack1 @gleckler1 these new tests have been run on crunchy with esmpy rather than esmp please check that the new values are reasonable.

### CODE REVIEW

**Comment by durack1** [comment]
@doutriaux1 what is the difference between the v1.0 and v2.0 jsons - are these the different formats related to the v1.1.1 release and our as yet updated provenance stuff?

**Comment by doutriaux1** [comment]
@durack1 yes `v1` were the very first one, `v2` was our first tweaks, we are currently on `v3` which includes the json_struct and sorts. The `JSONs` class is backward compatible and figures out which version of the json file is being passed to it.

**Comment by durack1** [comment]
@doutriaux1 with the provenance stuff in #403, can we add `mesalib` and maybe also the `LAPACK/BLAS` libraries as defined in https://github.com/numpy/numpy/issues/3912#issuecomment-26226437

**Comment by doutriaux1** [comment]
I added the `mesalib` and `esmpy` to the provenance in this PR. Not sure we care about `lapack` becuase anyway they come from `conda-forge` like `numpy`

**Comment by durack1** [comment]
@doutriaux1 yes, but `LAPACK/BLAS` are what numpy uses, and could be a source of difference between `linux` and `osx`.. Maybe.. I would imagine these libraries are `OS` provided, so numpy just uses whatever is available..?

**Comment by doutriaux1** [comment]
now i realize I didn't update the provenance in the jsons... Will upload a new one shortly.

File: tests/pcmdi/obsByNameTest/tas_2.5x2.5_regrid2_linear_metrics.json [hunk 1]
```diff
@@ -597,6 +597,46 @@
...
+            "Platform": "linux-64",
+            "Version": "4.3.11",
+            "IsPrivate": "False",
+            "envVersion": "4.3.11",
+            "buildVersion": "2.1.10",
+            "PythonVersion": "2.7.13.final.0",
+            "RootEnvironment": "/export/doutriaux1/miniconda2  (writable)",
+            "DefaultEnvironment": "/export/doutriaux1/miniconda2/envs/pmp"
+        },
+        "packages": {
```
Comment (durack1) [inline tests/pcmdi/obsByNameTest/tas_2.5x2.5_regrid2_linear_metrics.json]: There is no `mesalib` (or `LAPACK/BLAS`) in here

**Comment by doutriaux1** [comment]
@durack1 i will add them. nupmy actually does not use them! (see: https://github.com/conda-forge/numpy-feedstock/blob/master/recipe/meta.yaml) but we do for libcf I believe.

**Comment by durack1** [comment]
@doutriaux1 looks like both are in fact built: https://github.com/conda-forge/numpy-feedstock/blob/master/recipe/meta.yaml#L27-L28

**Comment by doutriaux1** [comment]
@durack1 that's blas not lapack

**Comment by durack1** [comment]
@doutriaux1 looks like the test uses `LAPACK` https://github.com/conda-forge/numpy-feedstock/blob/master/recipe/meta.yaml#L41

**Comment by doutriaux1** [comment]
that's numpy's own lapack. No dependencies on any conda lapack package.

**Comment by durack1** [comment]
@doutriaux1 numpy doesn't include `LAPACK`, it's an OS provided library isn't it? And if so, then the version of the library would differ across OSes.. Right, and could that then be a potential source of differences across `osx` vs `linux`, and between `linux` variants too? So Ubuntu is much more up-to-date that RHEL for e.g.

**Comment by doutriaux1** [comment]
@durack1 not it's not necessarily os provided. conda-forge provides them. I doubt it's our mac vs linux issue, but it might. Anyhow I now record them in the provenance.

**Comment by durack1** [comment]
@doutriaux1 great, I hope this might provide the additional info we need to sort out the osx vs linux issue.. It'd be a huge jump forward if we can resolve that.. I think it's almost worthy of a 2.0 tag just for that!

**Comment by doutriaux1** [comment]
@durack1 look at the mac resluts, while a bit off, the numbers are `very` close now.
https://travis-ci.org/PCMDI/pcmdi_metrics/jobs/241315892

**Comment by durack1** [comment]
@doutriaux1 great! The link above didn't show me a log, so not sure how close these are.. But convergence is a really good sign..

**Comment by doutriaux1** [comment]
@durack1 the log is in the link just scroll down enough you'll see it, one of the two failing tests is at the very bottom.

**Comment by durack1** [comment]
@doutriaux1 I tried again and saw:
```
Comparing: /Users/travis/build/PCMDI/pcmdi_metrics/pcmdi_install_test_results/metrics_results/salinityTest/sos_2.5x2.5_esmf_linear_metrics.json /Users/travis/build/PCMDI/pcmdi_metrics/tests/pcmdi/salinityTest/sos_2.5x2.5_esmf_linear_metrics.json atol: 0.01
Error for: GFDL-CM3,  JPL-Aquarius-v2,  r1i1p1,  NHEX,  mean-obs_xy,  djf,  ( 0, 0, 0, 0, 3, 1, ). Test value 34.222 vs expected value: 34.236
Error for: GFDL-CM3,  JPL-Aquarius-v2,  r1i1p1,  NHEX,  rms_xy,  djf,  ( 0, 0, 0, 0, 6, 1, ). Test value 0.893 vs expected value: 0.874
Error for: GFDL-CM3,  JPL-Aquarius-v2,  r1i1p1,  NHEX,  rmsc_xy,  djf,  ( 0, 0, 0, 0, 9, 1, ). Test value 0.839 vs expected value: 0.822
Error for: GFDL-CM3,  JPL-Aquarius-v2,  r1i1p1,  NHEX,  std-obs_xy,  djf,  ( 0, 0, 0, 0, 10, 1, ). Test value 1.586 vs expected value: 1.567
```
So yes, this is all second decimal place, which is good.. Trying to figure out why is the next (and harder) question.. Is this the same OS comparing the results - or osx vs linux?

**Comment by doutriaux1** [comment]
@durack1 `json` file have been generated on `crunchy` and the link is the results ran on osx, so it's osx vs linux. travis linux passes.

**Comment by durack1** [comment]
@doutriaux1 `BLAS` isn't something that you can get following the other packages, rather you have to use `np.__config__` which spits out the identical values for `osx` and `linux` for me:
```
UV-CDAT 2.8.0 OSX:
In [2]: np.__config__.show()
lapack_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/include']
blas_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/include']
openblas_lapack_info:
  NOT AVAILABLE
lapack_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/include']
blas_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/durack1/anaconda2/envs/uvcdat280/include']
```
```
UV-CDAT 2.10.0 linux/RHEL6.9:
In [2]: np.__config__.show()
lapack_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/include']
blas_opt_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/include']
lapack_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/include']
blas_mkl_info:
    libraries = ['mkl_intel_lp64', 'mkl_intel_thread', 'mkl_core', 'iomp5', 'pthread']
    library_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/export/durack1/anaconda2/envs/uvcdat2100nox/include']
```

**Comment by doutriaux1** [comment]
@durack1 it comes as a conda `feature` for numpy, so if you install it it automatically get you the matching `numpy`

**Comment by doutriaux1** [comment]
@zshaheen @durack1 @gleckler1 looking at this mac vs Lniux thing

climo maps for gensftlf (one of the failing test)
relative error (mac-linux).linux are:

 -5.70873499552e-16 6.25398454639e-16

So I don't think that's our issue.

I really don't know where the diff comes from. 

@dnadeau4 could you spend some time on this with me and @zshaheen tomorrow? That would be awesome.

**Comment by durack1** [comment]
@doutriaux1 yep `e-16` looks to me like a precision thing, so that's good news

**Comment by durack1** [comment]
@doutriaux1 @gleckler1 are you waiting for a review and merge here, or is there more work to be finalized?

**Comment by doutriaux1** [comment]
@gleckler1 reviewed the new number we are 99.9% sure it's good to go but awaiting total confirmation (some work with GFDL) before merging in.

**Comment by durack1** [comment]
@doutriaux1 great, I don't want to hold anything as important as this up so let me know when to review and merge this final product..

**Comment by zshaheen** [comment]
Just to confirm with you guys, we want to tie v1.1.2 with a specific version of CDAT correct? So not something like >= 2.10, instead all packages in meta.yaml would be equal to 2.10.

**Comment by durack1** [comment]
@zshaheen exactly, there have been no patch fixes to any of the sub-packages (e.g. cdms2, cdutil, genutil), so locking to v2.10 would be ideal - if we find there is a standing issue with the cdutil.ANNUAL function, then it would be good to fix this in cdutil and releasing so that we're not knowingly distributing errored software

**Comment by gleckler1** [comment]
@zshaheen confirmed.   That gives us a "locked-down" version, rather than something in a state of flux.

**Comment by durack1** [comment]
@doutriaux1 @zshaheen @gleckler1 woo hoo!

**Pull request merged**